service/worker/archiver/activities.go

```diff
@@ -117,18 +120,25 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
 		historyBlobReader = NewHistoryBlobReader(NewHistoryBlobIterator(request, container, domainName, clusterName))
 	}
 	blobstoreClient := container.Blobstore
+
+	// start uploading blobs
 	handledLastBlob := false
 	for pageToken := common.FirstBlobPageToken; !handledLastBlob; pageToken++ {
 		key, err := NewHistoryBlobKey(request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, pageToken)
 		if err != nil {
-			logger.Error(uploadErrorMsg, tag.UploadFailReason("could not construct blob key"), tag.ArchivalBucket(bucket))
-			return cadence.NewCustomError(errConstructBlob)
+			logger.Error(uploadErrorMsg, tag.UploadFailReason("could not construct blob key"), tag.ArchivalBucket(request.BucketName))
+			return cadence.NewCustomError(errConstructKey)
 		}
-		tags, err := getTags(ctx, blobstoreClient, bucket, key)
+
+		// get blob tag using constructed key to see if the blob exists and if it's the last blob
+		tags, err := getTags(ctx, blobstoreClient, request.BucketName, key)
 		if err != nil && err != blobstore.ErrBlobNotExists {
-			logger.Error(uploadErrorMsg, tag.UploadFailReason("could not get blob tags"), tag.ArchivalBucket(bucket), tag.ArchivalBlobKey(key.String()))
+			logger.Error(uploadErrorMsg, tag.UploadFailReason("could not get blob tags"), tag.ArchivalBucket(request.BucketName), tag.ArchivalBlobKey(key.String()))
 			return err
 		}
+
+		// if we can get tag, it means this blob already exists
+		// Based on if it's the last blob, we need to break the loop or continue
 		runConstTest := false
 		blobAlreadyExists := err == nil
```

Comment:
Here is an example of what I mean by making code clear enough it does not need comments. I could have written this code as `if err == nil`. But that would be confusing. So instead I created a named variable to describe what I mean when `err == nil` and my if uses that variable.

Reply:
Got it!